ksql-api/src/test/java/io/confluent/ksql/api/BufferedPublisherTest.java

```diff
@@ -0,0 +1,430 @@
+  }
+
+  @Test
+  public void shouldDeliverAllRecordsWhenMoreAreRequested() throws Exception {
+    loadPublisher(10);
+    shouldDeliver(15, 10);
+  }
+
+  @Test
+  public void shouldDeliverMoreThanMaxSendBatchSize() throws Exception {
+    int num = 2 * BufferedPublisher.SEND_MAX_BATCH_SIZE;
+    loadPublisher(num);
+    shouldDeliver(num, num);
+  }
+
+  @Test
+  public void shouldDeliverAllRequestingOneByOne() throws Exception {
+    loadPublisher(10);
+    TestSubscriber subscriber = new TestSubscriber(context) {
+
+      @Override
+      public synchronized void onSubscribe(final Subscription sub) {
+        super.onSubscribe(sub);
+        sub.request(1);
+      }
+
+      @Override
+      public synchronized void onNext(final String value) {
+        super.onNext(value);
+        getSub().request(1);
+      }
+    };
+    subscribeOnContext(subscriber);
+    assertThatEventually(subscriber::getValues, hasSize(10));
+    for (int i = 0; i < 10; i++) {
+      assertThat(subscriber.getValues().get(i), equalTo("record" + i));
+    }
+    assertThat(subscriber.isCompleted(), equalTo(false));
+    assertThat(subscriber.getError(), is(nullValue()));
+  }
+
+  @Test
+  public void shouldFailWhenZeroRequested() throws Exception {
+    loadPublisher(10);
+    TestSubscriber subscriber = new TestSubscriber(context) {
+
+      @Override
+      public synchronized void onSubscribe(final Subscription sub) {
+        super.onSubscribe(sub);
+        sub.request(0);
+      }
+    };
+    subscribeOnContext(subscriber);
+    assertThatEventually(subscriber::getError, is(notNullValue()));
+    assertThat(subscriber.getValues(), hasSize(0));
+    assertThat(subscriber.getError(), instanceOf(IllegalArgumentException.class));
+  }
+
+  @Test
+  public void shouldFailWhenNegativeRequested() throws Exception {
+    loadPublisher(10);
+    TestSubscriber subscriber = new TestSubscriber(context) {
+
+      @Override
+      public synchronized void onSubscribe(final Subscription sub) {
+        super.onSubscribe(sub);
+        sub.request(-1);
+      }
+    };
+    subscribeOnContext(subscriber);
+    assertThatEventually(subscriber::getError, is(notNullValue()));
+    assertThat(subscriber.getValues(), hasSize(0));
+    assertThat(subscriber.getError(), instanceOf(IllegalArgumentException.class));
+  }
+
+  @Test
+  public void shouldCompleteWhenNoRecords() throws Exception {
+    TestSubscriber subscriber = new TestSubscriber(context);
+    subscribeOnContext(subscriber);
+    execOnContextAndWait(publisher::complete);
+    assertThatEventually(subscriber::isCompleted, equalTo(true));
+  }
+
+  @Test
+  public void shouldCompleteAfterDeliveringRecords() throws Exception {
+    loadPublisher(10);
+    AsyncAssert asyncAssert = new AsyncAssert();
+    TestSubscriber subscriber = new TestSubscriber(context) {
+      @Override
+      public synchronized void onSubscribe(final Subscription sub) {
+        super.onSubscribe(sub);
+        sub.request(1);
+      }
```

Comment:
If you had `TestSubscriber`'s constructor take the initial request size you could simplify a lot of these tests.